### ISSUE TITLE
[BUG][Fuzzer][wrong-code] `T.annotate_restrict_buffers` silently emits `__restrict__` on an aliasing `*_handle` buffer, giving a wrong result

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#2596]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

0.1.11 (source pinned to tag `v0.1.11`, commit `cd37ed5fc35ae7a60a1277c8eb49028174ac51e6`).

### System information

NVIDIA H100 80GB HBM3 (`sm_90`), Python 3.12, torch 2.12.1+cu130 (CUDA 13.0), tilelang 0.1.11 (tag `v0.1.11` = `cd37ed5fc35ae7a60a1277c8eb49028174ac51e6`). The defect is in a target-independent transform + the CUDA codegen restrict logic; reproduced on `sm_90` this session.

### Problem description

When a buffer passed to `T.annotate_restrict_buffers` has a name ending in `_handle`, the pass silently drops it from the "do-not-restrict" set, so codegen emits `__restrict__` on it anyway — the exact aliasing guarantee the annotation exists to *suppress*. Under real aliasing this produces a silently wrong result: `nvcc` reuses a stale cached load on the false disjointness promise. A byte-identical kernel whose buffer is named `B` (no `_handle` suffix) is correct.

Concretely, for `annotate_restrict_buffers(A, A_handle)` where `A` and `A_handle` are aliased at the call site, `A_handle` is emitted as `float* __restrict__ A_handle`; the kernel then computes `D = 2*(i+1)` instead of the correct `7 + (i+1)`. The control naming the buffer `B` emits no `__restrict__` and computes the correct result.

<details>
<summary>Emitted signatures + runtime values (verified this session, sm_90/H100, deterministic x2)</summary>

| 2nd buffer name | emitted qualifier | `D[:4]` under alias | correct? |
|---|---|---|---|
| `A_handle` (bug) | `float* __restrict__ A_handle` | `[2, 4, 6, 8]` | ✗ (silent) |
| `B` (control) | `float* B` (no restrict) | `[8, 9, 10, 11]` | ✓ |

Reference under aliasing: `t=A[i]=i+1`, then `A_handle[i]=C[i]=7` (aliases `A[i]`), then `D[i]=A[i]+t = 7+(i+1)`. The wrong `[2,4,6,8]` is `2*(i+1)` — the stale `t` was reused as `A[i]`. The bug is order-dependent: the *second* buffer of the annotated pair whose normalized name collides is the one dropped.

</details>

### Reproducible example code

```python
import torch
import tilelang
import tilelang.language as T

N, DT = 256, "float32"


# BUG: the 2nd buffer is named `A_handle` (normalizes to `A`, colliding with `A`).
@T.prim_func
def bug(A: T.Tensor((N,), DT), A_handle: T.Tensor((N,), DT),
        C: T.Tensor((N,), DT), D: T.Tensor((N,), DT)):
    with T.Kernel(N, threads=1) as i:
        T.annotate_restrict_buffers(A, A_handle)   # declare A, A_handle MAY alias
        t = T.alloc_local((1,), DT)
        t[0] = A[i]
        A_handle[i] = C[i]                          # write through the aliased buffer
        D[i] = A[i] + t[0]                          # reload A[i]; under alias A[i]==7


# CONTROL: byte-identical but the 2nd buffer is named `B` (no `_handle` collision).
@T.prim_func
def control(A: T.Tensor((N,), DT), B: T.Tensor((N,), DT),
            C: T.Tensor((N,), DT), D: T.Tensor((N,), DT)):
    with T.Kernel(N, threads=1) as i:
        T.annotate_restrict_buffers(A, B)
        t = T.alloc_local((1,), DT)
        t[0] = A[i]
        B[i] = C[i]
        D[i] = A[i] + t[0]


def run(pf, second_name, tag):
    k = tilelang.compile(pf, target="cuda")
    sig = next(l for l in k.get_kernel_source().splitlines() if "__global__" in l)
    A = torch.arange(1, N + 1, device="cuda", dtype=torch.float32).contiguous()
    Cc = torch.full((N,), 7.0, device="cuda", dtype=torch.float32)
    D = torch.empty(N, device="cuda", dtype=torch.float32)
    k(A, A, Cc, D)                                  # pass A twice -> A and 2nd alias
    ref = 7.0 + torch.arange(1, N + 1, dtype=torch.float32)
    print(f"[{tag}] __restrict__ on 2nd: {f'__restrict__ {second_name}' in sig} | "
          f"D[:4]={D[:4].cpu().tolist()} correct={bool(torch.equal(D.cpu(), ref))}")


run(bug, "A_handle", "BUG")       # -> __restrict__ True  | D[:4]=[2,4,6,8]   correct=False
run(control, "B", "CONTROL")      # -> __restrict__ False | D[:4]=[8,9,10,11] correct=True
```

### Traceback

```pytb
No traceback — the kernel compiles and runs to completion; the result is silently wrong and deterministic.
```

### Expected behavior

Every buffer passed to `T.annotate_restrict_buffers` should be kept out of the `__restrict__` set regardless of its name — the annotation's whole purpose is to tell codegen these buffers may alias. The control (buffer named `B`) already does the right thing; a name ending in `_handle` should not change the result.

### Additional context

**Root cause.** [`HoistNonRestrictParams`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/hoist_nonrestrict_params.cc#L39-L58) collects the annotated buffers into the `kNonRestrictParams` set via `MaybeInsert`, which dedups by *pointer* **and** by a normalized name: [`NormalizeName` strips a trailing `_handle` (`hoist_nonrestrict_params.cc:39-44`)](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/hoist_nonrestrict_params.cc#L39-L44), and [`MaybeInsert` skips a buffer whose normalized name was already seen (`:53-55`)](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/transform/hoist_nonrestrict_params.cc#L46-L58). A user buffer literally named `A_handle` has data-var name `A_handle` → `NormalizeName` → `A`, which collides with buffer `A`'s data-var `A`, so the second is dropped from the set. Codegen then emits `__restrict__` on any no-alias parameter *not* in that set ([`codegen_cuda.cc:5476`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/cuda/codegen/codegen_cuda.cc#L5476) and [`:5585`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/cuda/codegen/codegen_cuda.cc#L5585)), so the dropped buffer is wrongly restricted. The `_handle` suffix is what TileLang's own frontend appends to handle vars, which is presumably why the normalizer strips it — but it collides with user buffers that end in `_handle`.

**Suggested fix.** The pointer-identity dedup (`seen_ptr_`) already prevents double-insertion; the name-based dedup (`seen_name_` / `NormalizeName`) is what introduces the false collision. Dropping the normalized-name path, or normalizing on the parameter name rather than the buffer data-var name, would keep every annotated buffer in the set.

**Provenance.** `NormalizeName` and the `MaybeInsert` name-dedup were introduced with the file in [#1428](https://github.com/tile-ai/tilelang/pull/1428) ("[Language] Introduce `T.annotate_restrict_buffers`", merged 2025-12-17). Present since the feature first shipped — not a regression.

I searched the open and closed tracker for `annotate_restrict` / `HoistNonRestrict` / `non_restrict` / `__restrict__` and found no existing report of this defect.
